# HaplowebMaker and CoMa
HaplowebMaker creates automatically haplowebs (Flot et al. 2010) starting from FASTA files. CoMa combines the informations from several markers into a single conspecificity matrix (Debortoli et al. 2016).

# Installation

Just copy / paste this folder to a webserver.

If you want to use HaplowebMaker offline, either install a local webserver (recommended) or activate CORS requests so that the HaplowebMaker website can load it's resources in the corresponding subdirectories
(This is not recommended as this has some security implications).

In order to activate CORS requests in Firefox, enter about:config as URL, accept the warning message and set the setting "privacy.file_unique_origin" to false.

# Usage

Open http(s)://&lt;name of your webbrowser&gt;/&lt;directory&gt;/index.html in any modern browser that fully supports HTML 5.

# Relevant papers
Bandelt, Forster & RÃ¶hl (1999) Median-joining networks for inferring intraspecific phylogenies. Molecular biology and evolution 16:37-48

Flot, Couloux & Tillier (2016) "Haplowebs as a graphical tool for delimiting species: a revival of Doyle's" field for recombination" approach and its application to the coral genus Pocillopora in Clipperton." BMC Evolutionary Biology 10:372

Debortoli, Li, Eyres, Fontaneto, Hespeels, Tang, Flot, Van Doninck (2016) Genetic exchange among bdelloid rotifers is more likely due to horizontal gene transfer than to meiotic sex. Current Biology 26:723-732


